alpa/serialization.py

```diff
@@ -28,169 +25,137 @@
     """Save a checkpoint of the `target` to `path`. 
 
         Similar to flax.training.checkpoints.save_checkpoint, but support DistributedArrays 
-        and ReplicatedDistributedArray in alpa. Also it will save np.ndarray and jax.xla.DeviceArray
-        into tensorstore for later distributed loading.
+        and ReplicatedDistributedArray in alpa. 
         # TODO (zhongyinmin): copy all the safe-saving stuff from 
         https://flax.readthedocs.io/en/latest/_modules/flax/training/checkpoints.html#save_checkpoint
 
         Args:
-           ckpt_dir: str or pathlib-like path to store checkpoint directories in.
-           target: serializable flax object, usually a trainState
+           nfs_dir: shared filesystem path to store checkpoint. 
+           target: serializable flax object, usually a trainState.
            step: training step number or other metric number
+           local_cache_dir: If not None, `save_checkpoint` will return immediately after each host saving
+           its part of the model into this cache directory, and only the metadata file will be saved into
+           the `nfs_dir`. You can then use `rsync` to easily merge these these local data into `nfs_dir`.
     """
-    state_dict = to_state_dict(target)
-    os.makedirs(ckpt_dir, exist_ok=True)
-    ckpt_path = os.path.join(ckpt_dir, f"checkpoint_{step}")
-    with open(ckpt_path, 'wb') as fp:
-        fp.write(
-            msgpack.packb(state_dict,
-                          default=_msgpack_ext_pack_wrapper(ckpt_dir),
-                          strict_types=True))
+    # create directories if not exist
+    os.makedirs(nfs_dir, exist_ok=True)
+
+    if local_cache_dir is not None:
+        os.makedirs(local_cache_dir, exist_ok=True)
+        save_dir = local_cache_dir
+    else:
+        save_dir = nfs_dir
+
+    target = to_state_dict(target)
+    flat_dirs = _dfs_pytree(target, "state")
+    flat_target, target_tree = tree_flatten(target)
+    flat_metadata = []
+    obj_refs = []
+    assert(len(flat_dirs) == len(flat_target))
+    for arr_dir, x in zip(flat_dirs, flat_target):
+        if isinstance(x, (DistributedArray, ReplicatedDistributedArray, np.ndarray, jax.xla.DeviceArray)):
+            if isinstance(x, DistributedArray):
+                obj_refs.extend(x.save(os.path.join(save_dir, arr_dir), False))
+            elif isinstance(x, ReplicatedDistributedArray):
+                obj_refs.extend(x.replica.save(os.path.join(save_dir, arr_dir), False))
+            elif isinstance(x, (np.ndarray, jax.xla.DeviceArray)):
+                _save_unsharded_array(os.path.join(save_dir, arr_dir), x)
+            flat_metadata.append(arr_dir)
+        else:
+            flat_metadata.append(x)
+
+    metapath = os.path.join(nfs_dir, f"checkpoint_{step}")
+    metadata = tree_unflatten(target_tree, flat_metadata)
+    with open(metapath, "wb") as metafile:
+        metafile.write(msgpack.packb(metadata))
+    ray.get(obj_refs)
```

Comment:
You do not need to call `ray.get` here. All tests should still be correct. Users can explicitly call sync outside this function by `executable.sync()`, `device_mesh.sync_workers()` or `array.block_until_ready()`